common/scala/src/main/scala/whisk/common/tracing/OpenTracingProvider.scala

```diff
@@ -0,0 +1,220 @@
+            contextMap = Some(map.asScala.toMap)
+          }
+          case _ =>
+        }
+      }
+      case None => None
+    }
+    contextMap
+  }
+
+  /**
+   * Get the current TraceContext which can be used for downstream services
+   *
+   * @param transactionId
+   * @return
+   */
+  override def setTraceContext(transactionId: TransactionId, context: Option[Map[String, String]]) = {
+    context match {
+      case Some(scalaMap) => {
+        var javaMap: java.util.Map[String, String] =
+          scala.collection.JavaConverters.mapAsJavaMapConverter(scalaMap).asJava
+        var ctx: SpanContext = tracer.extract(Format.Builtin.TEXT_MAP, new TextMapExtractAdapter(javaMap))
+        TracingCacheProvider.contextCache.put(transactionId.meta.id, ctx)
+      }
+      case None =>
+    }
+  }
+
+  private def clear(transactionId: TransactionId): Unit = {
+    TracingCacheProvider.spanCache.get(transactionId.meta.id) match {
+      case Some(spanList) => {
+        spanList.last.get.map(_.finish)
+        spanList.remove(spanList.size - 1)
+        if (spanList.isEmpty) {
+          TracingCacheProvider.spanCache.invalidate(transactionId.meta.id)
+          TracingCacheProvider.contextCache.invalidate(transactionId.meta.id)
+        }
+
+      }
+      case None =>
+    }
+  }
+}
+
+trait WhiskTracer {
+  def startSpan(logMarker: LogMarkerToken, transactionId: TransactionId): Unit = {}
+  def finishSpan(transactionId: TransactionId): Unit = {}
+  def error(transactionId: TransactionId): Unit = {}
+  def getTraceContext(transactionId: TransactionId): Option[Map[String, String]] = None
+  def setTraceContext(transactionId: TransactionId, context: Option[Map[String, String]]): Unit = {}
+}
+
+object WhiskTracerProvider {
+  val tracingConfig = loadConfigOrThrow[TracingConfig](ConfigKeys.tracing)
+
+  val tracer: WhiskTracer = createTracer(tracingConfig)
+
+  private def createTracer(tracingConfig: TracingConfig): WhiskTracer = {
+
+    tracingConfig.zipkin match {
+      case Some(zipkinConfig) => {
+        val zipkinConfig = tracingConfig.zipkin.get
+        if (!GlobalTracer.isRegistered) {
+          val sender: Sender = OkHttpSender.create(zipkinConfig.getUrl)
+          val spanReporter = AsyncReporter.create(sender)
+          val braveTracing = Tracing
+            .newBuilder()
+            .localServiceName(tracingConfig.component)
+            .spanReporter(spanReporter)
+            .sampler(Sampler.create(zipkinConfig.sampleRate.toFloat))
+            .build();
+
+          //register with OpenTracing
+          GlobalTracer.register(BraveTracer.create(braveTracing))
+        }
+      }
+      case None =>
+    }
+
+    if (GlobalTracer.isRegistered)
+      new OpenTracer(GlobalTracer.get())
+    else
+      NoopTracer
+  }
+}
+
+private object NoopTracer extends WhiskTracer
+case class TracingConfig(component: String, cacheExpiry: Option[Int] = None, zipkin: Option[ZipkinConfig] = None)
```

Comment:
Instead of using `Option[Int]` prefer using `Duration` then config can have values like `1 minute`. No need for making it `Option` as default value is specified in `application.conf`.